### PR TITLE
Parsing Code39 barcodes. Methods to get AIM identifier and modifier

### PR DESCRIPTION
**Pull request by neemevool**
Parsing Code39 barcodes. Recognizes and interprets barcode scanner AIM modifiers for Code39. AimParser introduces methods to get the AIM identifier and identifier modifier from the raw reading. Possibility to extend other barcode types by reader modifier.

### About AIM identifier

AIM identifier (now actually ISO/IEC 15424:2008(E) standard) specifies a preamble message generated by the reader and interpretable by the receiving system, which indicates the bar code symbology or other origin of transmitted data, together with details of certain specified optional processing features associated with the data message. 

The AIM identifer is 3 character preamble in the reading, where ] is the "flag character" and indicates to the host that it and the characters following are the symbology identifier characters. Code character is second character in the symbology identifier string, which usually indicates to the host the bar code symbology of the symbol which has been read. Modifier character(s) are one or more characters following the code character in the symbology identifier string, indicating optional features or processing applied to the symbol.

The precise interpretation of the modifier character should be obtained by reference to the relevant symbology specification. The modifier characters define the options available for the code character. The number of modifier characters and their meaning is defined for each of the code characters. The first modifier character shall be from the set {0 to 9, A to Z, a to z}; in some instances the character may represent a hexadecimal value (0 to F) corresponding to the sum of active processing options.

### CODE REVIEW

**Comment by DacoTaco** [comment]
hey, thanks for the emails/Pull Request!

generally, this is an awesome improvement that i think is very good to have!
this would also improve the GS1(28) parsing, incase the FUNC1 is defined differently in the scanner.

however, i have a general question why it would be important to have the `AimReaderModifier` available in the barcode. once you have the barcode, this information is no longer relevant, correct?

can't we have this info be passed to the parse function of the barcodeparserbuilder, so that when its parsing it can use it to parse the barcode in a specific way and/or save it to the specific barcode *ïf* needed ?

File: BarcodeParserBuilder/BarcodeParserBuilder.cs [hunk 1]
```diff
@@ -35,6 +35,7 @@ public bool TryParse(string? barcodeString, out Barcode? barcode, out string? fe
 
                     //retrieve output parameter and return true
                     barcode = (Barcode?)tryParseParameters[1];
+                    feedback = $"was parsed by {parserBuilder.Name}";
```
Comment (DacoTaco) [inline BarcodeParserBuilder/BarcodeParserBuilder.cs:38]: why are you setting the feedback? don't you know what parser builder parsed it based on the type of barcode youre getting back?

File: BarcodeParserBuilder/Infrastructure/AimReaderModifier.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                
+
+        public override bool Equals(object obj)
+        {
+            return obj is AimReaderModifier aimModifier && aimModifier.Value == this.Value;
+        }
+
+        public int CompareTo(object other) => Value.CompareTo(((AimReaderModifier)other).Value);
+
+        //public AimReaderModifier(string readerModifierValue) { Value = readerModifierValue ?? throw new ArgumentNullException(nameof(readerModifierValue)); }
```
Comment (DacoTaco) [inline BarcodeParserBuilder/Infrastructure/AimReaderModifier.cs:35]: commented code -> delete

File: BarcodeParserBuilder/Infrastructure/AimParser.cs [hunk 3]
```diff
@@ -59,6 +60,31 @@ internal static string StripBarcodePrefix(string barcodeString)
...
+        {
+            return barcodeString.Length > 3 && barcodeString.StartsWith("]");
+        }
+
+        /// <summary>
+        /// AIM identifier consist of code and modifier. AIM code defines the barcode type, for example EAN, CODE39, GS1 etc
+        /// </summary>
+        /// <param name="barcodeString">raw string from the reader</param>
+        /// <returns></returns>
+        internal static string GetAimIdentifierCode(string barcodeString)
```
Comment (DacoTaco) [inline BarcodeParserBuilder/Infrastructure/AimParser.cs:73]: why not create a class called AimInformation and just return that info in 1 go? no need to call both functions if you always want the full information anyway

File: BarcodeParserBuilder/Barcodes/CODE39/Code39Barcode.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        if (Code39ReaderModifier.GetAllValues<Code39ReaderModifier>().Contains(readerModifierValue))
+        {
+            return new Code39ReaderModifier(readerModifierValue);
+        }
+
+        throw new Code39ParseException("Invalid reader modifier");
+    }
+
+    public static string StripCheckCharacter(string inputString, Code39ReaderModifier readerModifier)
```
Comment (DacoTaco) [inline BarcodeParserBuilder/Barcodes/CODE39/Code39Barcode.cs:71]: what is the use of this?

**Review by DacoTaco — CHANGES_REQUESTED** [review]
overal great addition. just some comments i'd like to see changed :) ^^

**Comment by neemevool** [comment]
Hello!

 

Yes, I considered it to _not_ pass the modifier but the problem is that it may be needed when using the barcode data. 

This is because how real scanners work. Scanners have for each barcode type configurable modifier options and these affect the final outcome. 

For example the same A – Code39. It is possible to tell the scanner just alone for that Code39 settings:

*	Please do full ASCII conversion for me
*	Please don’t do any conversion
*	Please calculate the checksum and pass it as part of the value
*	Please don’t do checksum validation
*	Please validate the checksum and do _not_ pass any barcode when it does not contain checksum

 

Sometimes software developer can, sometimes cannot dictate what are the scanner settings applied. So in some circumstances you must be able to strip the checksum yourself, sometimes you want to have it not passed at all. 

 

It gets even more complex for other barcode types. So thereby I thought that this information would be needed to be passed with the reading.

**Comment by neemevool** [comment]
Yes, I will do changes. But please view some questions below

**Review by DacoTaco — APPROVED** [review]

**Pull request merged**